### PR TITLE
Use ruff for linting

### PR DESCRIPTION
**Pull request by maxnoe**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -69,3 +69,20 @@ norecursedirs = [
     [[tool.towncrier.section]]
         name = ""
         path = ""
+
+
+[tool.ruff]
+target-version = "py39"
+
+[tool.ruff.lint]
+select = ["E", "F", "I"]
```
Comment (kosack) [inline pyproject.toml:78]: I suggest adding a few here (perhaps in a separate PR if they are too strict):
- `NPY` (numpy rules)
- `PTH` (flake8-use-pathlib)
-  `UP` (pyupgrade)

too much for now, but maybe in the future...:
- `PERF` (performance checks) - currently finds 12 suggestions
-  `SIM` (simplify code)
-  `T20` (print statement), but exclude examples and the ctapipe-info tool
-  `D` (pydocstyle) - we have a  lot of issues here.

Comment (maxnoe) [reply pyproject.toml:78]: Yes, I planned to add iteratively. Pydocstyle as well

File: pyproject.toml [hunk 2]
```diff
@@ -7,7 +7,7 @@ write_to = "ctapipe/_version.py"
 
 [tool.black]
 line-length = 88
-target-version = ['py37', 'py38', 'py39', 'py310']
+target-version = ['py39', 'py310', 'py311']
 include = '\.pyi?$'
 exclude = '''
 /(
```
Comment (morcuended) [inline pyproject.toml:13]: do we want to keep black settings once ruff is in place?

File: docs/conf.py [hunk 3]
```diff
@@ -25,11 +25,8 @@
 # -- General configuration ------------------------------------------------
 import datetime
 import os
-
-# Get configuration information from setup.cfg
 from configparser import ConfigParser
 
-# Sphinx gallery
```
Comment (Tobychev) [inline docs/conf.py:32]: How come this comment got cut?

Comment (maxnoe) [reply docs/conf.py:36]: I found a comment "Sphinx gallery" before an import "sphinx_gallery" pretty useless...

File: src/ctapipe/io/__init__.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 # order matters to prevent circular imports
 isort:skip_file
 """
-from .astropy_helpers import read_table, write_table
+from .astropy_helpers import read_table, write_table  # noqa: I001
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply src/ctapipe/io/__init__.py:7]: The explanation is in the module docstring above.

The order is *not* alphabetical, the `datawriter` comes large, it depends on the other modules.

File: examples/core/table_writer_reader.py [hunk 5]
```diff
@@ -94,16 +89,14 @@ def create_stream(n_event):
...
         for data in create_stream(10):
-
             h5_table.write("table", data)
             0 / 0
 except Exception as err:
     print("FAILED:", err)
 print("Done")
 
-h5_table.h5file.isopen == False
+h5_table.h5file.isopen
```
[the remark this answers is not in the document]
Comment (maxnoe) [reply examples/core/table_writer_reader.py:99]: This is not a test but an example notebook (the output cell is shown to the user)

**Review by Tobychev — APPROVED** [review]

**Review by kosack — APPROVED** [review]

**Pull request merged**